---
title: "BaSiC"
author: "Wei Sun"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This R markdown file illustrate the usage of R package ```basic``` with application on the bulk sample and single cell sequencing data generated by Gawad, C., Koh, W., & Quake, S. R. (2014). *Dissecting the clonal origins of childhood acute lymphoblastic leukemia by single-cell genomics*. PNAS, 111(50), 17947-17952. This study collected bulk and single cell sequencing data from six childhood acute lymphoblastic leukemia (ALL) patients. This is a type of cancer with relatively small number of somatic mutations. Altogether 278 mutations are identified from 6 patients, with the number of mutations per patient varies from 10 to 105. After identifying somatic mutations from exome-seq of bulk samples, they sequenced 1,479 single tumor cells from these six patients using target sequencing with high depth. 

BaSiC is designed to analyze exome-seq data with a relatively large number of mutations per sample. In addition, Basic exploit the read count information to impute mutation calls when sequence depth in single cells are relatively low. Since this dataset has relatively small number of mutations and high read-depth in single cells, it is not the type data that BaSiC is designed for. Neverthless, we apply BaSiC to demonstrate its performance in this dataset. 

```{r libraries, warning = FALSE, message = FALSE}
library(basic)
library(gridExtra)
library(ggplot2)
library(ggcorrplot)
library(reshape2)
library(ggpubr)
```

We downloaded the information of somatic mutations of bulk tumor samples from [Supplementary Dataset_S01](https://www.pnas.org/highwire/filestream/617852/field_highwire_adjunct_files/1/pnas.1420822111.sd01.xlsx). This file provides mutation location and VAF, but does not provide read-depth information. We downloaded the raw sequence data (fastq files) from NCBI SRA, mapped the reads to hg19 using BWA with default options, and used the following script to process these raw data to obtain read coutn information. 

```
#!/bin/bash 
java -Xmx24g -jar path2Picard/picard.jar SortSam \
      INPUT=SRR1517762_hg19.sam \
      OUTPUT=SRR1517762_hg19_sorted.bam \
      SORT_ORDER=coordinate 
      
java -Xmx24g -jar path2Picard/picard.jar AddOrReplaceReadGroups \
      INPUT=SRR1517762_hg19_sorted.bam \
      OUTPUT=SRR1517762_hg19_sorted_RG.bam \
      ID=FLOWCELL6.LANE6\
      SORT_ORDER=coordinate \
      LB=SRR1517762 PL=ILLUMINA \
      SM=SRR1517762 \
      PU=FLOWCELL6.LANE6 
      
java -Xmx24g -jar path2Picard/picard.jar MarkDuplicates REMOVE_DUPLICATES=true \
      INPUT=SRR1517762_hg19_sorted_RG.bam \
      OUTPUT=SRR1517762_hg19_sorted_RG_dedup.bam \
      METRICS_FILE=SRR1517762_hg19_sorted_RG_dedup_metric.txt 
      
samtools index SRR1517762_hg19_sorted_RG_dedup.bam 

java -Xmx7g -jar GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar \
      -T ASEReadCounter \
      -R /fh/fast/sun_w/research/data/human/hg19/Homo_sapiens_assembly19.fasta \
      -I SRR1517762_hg19_sorted_RG_dedup.bam \
      -o ./output/SRR1517762_hg19_site2.txt \
      -sites vcf_sites2.vcf \
      -minDepth 1 \
      --minMappingQuality 1 \
      --minBaseQuality 0 \
      --disable_auto_index_creation_and_locking_when_reading_rods
```

First read in the somatic mutaton calls and read-depth information from bulk tumor. 
```{r load_bulk_data}

mutBk = read.table("basic/data/Final_Mutation_Calls_with_counts", sep="\t", 
                  as.is=TRUE, header=TRUE)
dim(mutBk)
mutBk[1:2,]
mutBk$loci = paste(mutBk$Chromosome, mutBk$Position, sep="_")

table(mutBk$Patient, useNA="ifany")
table(mutBk$Type, mutBk$Effect, useNA="ifany")
```

Note that the VAF calcualted from our read counts are strongly correlated with, but different from ```Confirmed.Tumor.Var.Freq``` reported by the authors. This is because the ```Confirmed.Tumor.Var.Freq``` was obtained from target sequncing validations. 

Next we read in the single cell data
```{r load_sc_data}
mutS = clustS = gL = list()

for(i in 1:6){
  fname = sprintf("basic/data/pat_%d_cluster_dat", i)
  mut.i = read.table(fname, header=TRUE)
  mut.i = data.matrix(mut.i)
  t1    = mut.i[,which(colnames(mut.i) != "clusters")]
  clustS[[i]] = mut.i[,which(colnames(mut.i) == "clusters")]

  h1 = hclust(dist(t1))
  h2 = hclust(dist(t(t1)))
  t1 = t1[h1$order, h2$order]
  mutS[[i]] = t1
}
```

The following codes show the dimension of the data from 6 patients (the 1st row is the number of cells and the 2nd row is the number of mutations), as well as the number of 0's (no mutation) and 1's (with mutation) in the mutation data matrix. 
```{r check_missing}
sapply(mutS, dim)
sapply(mutS, table, useNA="ifany")
```


Next we cluster the mutations based on the data from single cells or both bulk tumor and single cells. Since there is no missing values, the read count data are not needed. We simply set both trec and asrec to be matrix of 0's. We evaluate the results with number of clusters being 2, 3, 4, or 5. We skip patient 6 in our anlaysis becaue it appears all the 10 mutations of this patient are clonal. 

```{r analaysis, fig.height=4, fig.width=4}

mutB = gamma.plot.scs = gamma.plot.bsc = list()

for(i in 1:5){
  print(i)
  mutations = t(mutS[[i]])
  
  asrec  = trec = matrix(0, nrow=nrow(mutations), ncol=ncol(mutations))
  mutBki = mutBk[which(mutBk$Patient == paste("Patient", i)),]
  
  matchi = match(rownames(mutations), mutBki$loci)
  if(any(is.na(matchi))){ stop("some mutations were not found\n") }
  
  b.asrec = mutBki$altCount[matchi]
  b.trec  = mutBki$totalCount[matchi]
  b.cn    = rep(2, length(b.asrec))
  mutB[[i]] = cbind(b.asrec, b.trec)
  
  scs = cluster.SCS(mutations, asrec, trec, nclust=2:5, verbose=0)
  bsc = cluster.BSC(b.cn, b.asrec, b.trec, mutations, asrec, trec, 
                    nclusters=2:5, verbose=0)
  
  print(scs$summary.cluster)
  print(bsc$summary.cluster)
  
  cluster.scs = apply(scs[[scs$nclust.by.AIC]]$postPs, 1, which.max)
  cluster.bsc = apply(bsc[[bsc$nclust.by.AIC]]$postPs, 1, which.max)

  print(table(cluster.scs, cluster.bsc))
  
  if(any(order(cluster.bsc) != 1:nrow(mutations))){
    mutations = mutations[order(cluster.bsc),]
    mutB[[i]] = mutB[[i]][order(cluster.bsc),]
    par(mar=c(5,4,1,1), bty="n")
    plot(1:nrow(mutations), order(cluster.bsc), xlab="original order", 
         ylab="new order", pch=20, cex=0.5, col=rgb(0.8, 0.1, 0.1, 0.6))
  }
  
  gamma2plot = scs[[scs$nclust.by.AIC]]$gamma
  
  g0 = ggplot(melt(gamma2plot), aes(Var1, Var2, fill=value)) + 
    geom_tile(aes(fill = value), color = "white", size=0.25) + 
    scale_fill_gradient(low="white", high="red") +
    ylab("clusters") + xlab("cells") +  
    scale_x_discrete(expand = c(0, 0)) + 
    scale_y_discrete(expand = c(0, 0)) + 
    theme(axis.text.x=element_blank(), legend.position = "none") + 
    ggtitle('SCS only') + theme(plot.title = element_text(size = 10))
  
  gamma2plot = bsc[[bsc$nclust.by.AIC]]$gamma
  g1 = ggplot(melt(gamma2plot), aes(Var1, Var2, fill=value)) + 
    geom_tile(aes(fill = value), color = "white", size=0.25) + 
    scale_fill_gradient(low="white", high="red") +
    ylab("clusters") + xlab("cells") +  
    scale_x_discrete(expand = c(0, 0)) + 
    scale_y_discrete(expand = c(0, 0)) + 
    theme(axis.text.x=element_blank(), legend.position = "none") + 
    ggtitle('BaSiC') + theme(plot.title = element_text(size = 10))
  
  gamma.plot.scs[[i]] = g0
  gamma.plot.bsc[[i]] = g1

}
```

The following codes prepare the plots for SCS data as well as VAF from bulk tumor sequencing data. 
```{r plot0}
ga.plot = list()

for(i in 1:5){
  t2 = melt(mutS[[i]])
  names(t2)[1:2] = c("cell", "mutation")
  
  col2use = c("white", "darkred")
  
  g1 = ggplot(t2, aes(cell, mutation)) + theme_minimal() +
    geom_tile(aes(fill = as.factor(value)), color = "white", size=0.25) +
    ylab("mutations") + xlab("cells") + font("y.text", size = 6) + 
    scale_fill_manual(values=col2use) +  
    theme(axis.text.x=element_text(size=10), legend.position = "none")
  
  mutBi = mutB[[i]]
  df1 = data.frame(mutation=1:nrow(mutBi), VAF=mutBi[,1]/mutBi[,2])
  g2  = ggplot(data=df1, aes(x=mutation, y=VAF)) + 
        scale_x_discrete(expand = c(0, 0)) +
        geom_bar(stat="identity", fill="steelblue") + theme_minimal()
  g2  = g2 + coord_flip()

  ga = ggarrange(g1, g2, ncol=2, align = "h", widths = c(2, 1))
  ga.plot[[i]] = ga
}
```

Finally we illustrate the data and estimation results. 

```{r plot1a, fig.height=2.0, fig.width=7.5, echo = FALSE}
print(ga.plot[[1]])
```

```{r plot1b, fig.height=0.75, fig.width=5, echo = FALSE}
print(gamma.plot.scs[[1]])
print(gamma.plot.bsc[[1]])
```

```{r plot2a, fig.height=1.8, fig.width=7.5, echo = FALSE}
print(ga.plot[[2]])
```

```{r plot2b, fig.height=0.85, fig.width=5, echo = FALSE}
print(gamma.plot.scs[[2]])
print(gamma.plot.bsc[[2]])
```

```{r plot3a, fig.height=4.0, fig.width=9, echo = FALSE}
print(ga.plot[[3]])
```

```{r plot3b, fig.height=0.85, fig.width=6, echo = FALSE}
print(gamma.plot.scs[[3]])
print(gamma.plot.bsc[[3]])
```

```{r plot4a, fig.height=6.0, fig.width=9, echo = FALSE}
print(ga.plot[[4]])
```

```{r plot4b, fig.height=0.85, fig.width=6, echo = FALSE}
print(gamma.plot.scs[[4]])
print(gamma.plot.bsc[[4]])
```

```{r plot5a, fig.height=6.5, fig.width=9, echo = FALSE}
print(ga.plot[[4]])
```

```{r plot5b, fig.height=0.75, fig.width=6, echo = FALSE}
print(gamma.plot.scs[[4]])
print(gamma.plot.bsc[[4]])
```

```{r}
sessionInfo()
```
