---
title: "BaSiC"
author: "Wei Sun"
date: "1/7/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This R markdown file illustrate the usage of R package ```basic``` with application on the bulk smaple and single cell sequencing data generated by Gawad, C., Koh, W., & Quake, S. R. (2014). *Dissecting the clonal origins of childhood acute lymphoblastic leukemia by single-cell genomics*. PNAS, 111(50), 17947-17952. This study collected bulk and single cell sequencing data from six childhood acute lymphoblastic leukemia (ALL) patients. This is a type of cancer with relatively small number of somatic mutations. Altogether 278 mutations are identified from 6 patients, with the number of mutations per patient varies from 10 to 105. After identifing somatic mutations from exome-seq of bulk samples, they sequenced 1,479 single tumor cells from these six patients using target sequencing with high depth. 

BaSiC is designed to analyze exome-seq data with a relatively large number of mutations per sample. In addition, Basic exploit the read count information to impute mutation calls when sequence depth in single cells are relatively low. Since this dataset has relatively small number of mutations and high read-depth in single cells, it is not the type data that BaSiC is designed for. Neverthless, we apply BaSiC to demonstrate its performance in this dataset. 

```{r libraries, warning = FALSE, message = FALSE}
library(basic)
library(gridExtra)
library(ggplot2)
library(ggcorrplot)
library(reshape2)
```

We downloaded the information of somatic mutations of bulk tumor samples from [Supplementary Dataset_S01](https://www.pnas.org/highwire/filestream/617852/field_highwire_adjunct_files/1/pnas.1420822111.sd01.xlsx). This file provides mutation location and VAF, but does not provide read-depth information. We downloaded the raw sequence data (fastq files) from NCBI SRA, mapped the reads to hg19 using BWA with default options, and used the following script to process these raw data to obtain read coutn information. 

```
#!/bin/bash 
java -Xmx24g -jar path2Picard/picard.jar SortSam \
      INPUT=SRR1517762_hg19.sam \
      OUTPUT=SRR1517762_hg19_sorted.bam \
      SORT_ORDER=coordinate 
      
java -Xmx24g -jar path2Picard/picard.jar AddOrReplaceReadGroups \
      INPUT=SRR1517762_hg19_sorted.bam \
      OUTPUT=SRR1517762_hg19_sorted_RG.bam \
      ID=FLOWCELL6.LANE6\
      SORT_ORDER=coordinate \
      LB=SRR1517762 PL=ILLUMINA \
      SM=SRR1517762 \
      PU=FLOWCELL6.LANE6 
      
java -Xmx24g -jar path2Picard/picard.jar MarkDuplicates REMOVE_DUPLICATES=true \
      INPUT=SRR1517762_hg19_sorted_RG.bam \
      OUTPUT=SRR1517762_hg19_sorted_RG_dedup.bam \
      METRICS_FILE=SRR1517762_hg19_sorted_RG_dedup_metric.txt 
      
samtools index SRR1517762_hg19_sorted_RG_dedup.bam 

java -Xmx7g -jar GenomeAnalysisTK-3.6/GenomeAnalysisTK.jar \
      -T ASEReadCounter \
      -R /fh/fast/sun_w/research/data/human/hg19/Homo_sapiens_assembly19.fasta \
      -I SRR1517762_hg19_sorted_RG_dedup.bam \
      -o ./output/SRR1517762_hg19_site2.txt \
      -sites vcf_sites2.vcf \
      -minDepth 1 \
      --minMappingQuality 1 \
      --minBaseQuality 0 \
      --disable_auto_index_creation_and_locking_when_reading_rods
```

First read in the somatic mutaton calls and read-depth information from bulk tumor
```{r load_bulk_data}
mutBk = read.table("data/Final_Mutation_Calls_with_counts.txt", sep="\t", 
                  as.is=TRUE, header=TRUE)
dim(mutBk)
mutBk[1:2,]

table(mutBk$Patient, useNA="ifany")
table(mutBk$Type, mutBk$Effect, useNA="ifany")

plot(mutBk$Confirmed.Tumor.Var.Freq, mutBk$altCount/mutBk$totalCount)
```

Next we read in the single cell data
```{r load_sc_data, fig.height=4, fig.width=5}
mutS = clustS = list()

for(i in 1:6){
  fname = sprintf("data/pat_%d_cluster_dat", i)
  mut.i = read.table(fname, header=TRUE)
  mut.i = data.matrix(mut.i)
  t1    = mut.i[,which(colnames(mut.i) != "clusters")]
  clustS[[i]] = mut.i[,which(colnames(mut.i) == "clusters")]

  h1 = hclust(dist(t1))
  h2 = hclust(dist(t(t1)))
  t1 = t1[h1$order, h2$order]
  mutS[[i]] = t1
  rownames(t1) = 1:nrow(t1)
  colnames(t1) = 1:ncol(t1)
  
  t2 = melt(t1)
  names(t2)[1:2] = c("cell", "mutation")
  
  col2use = c("white", "darkred")
  
  g1 = ggplot(t2, aes(cell, mutation)) + theme_grey() +
    geom_tile(aes(fill = as.factor(value)), color = "white", size=0.25) +
    ylab("mutations") + xlab("cells") + 
    scale_x_discrete(expand = c(0, 0)) + 
    scale_y_discrete(expand = c(0, 0)) + 
    scale_fill_manual(values=col2use) +  
    theme(axis.text.x=element_blank(), legend.position = "none")
  print(sprintf("patient %d", i))
  print(g1)
}
```



```{r}
sessionInfo()
```
