---
title: "BaSiC"
author: "Wei Sun"
date: "1/3/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This R markdown file illustrate the usage of R package ```basic``` with application on the bulk smaple and single cell sequencing data generated by Wang et al. (2014) *Clonal evolution in breast cancer revealed by single nucleus genome sequencing*. Nature, 512(7513), 155-160.


```{r libraries, warning = FALSE, message = FALSE}
library(basic)
library(ggcorrplot)
library(gridExtra)
```

First read in the somatic mutaton calls and read-depth information from bulk tumor and single cells. 
```{r load_data}
load("data/wang_2014.RData")
ls()

dim(mutations.bulk)
mutations.bulk[1:2,]

dim(mutations.sc)
mutations.sc[1:2,]
```

We have filtered the somatic mutations so that the copy number is either 2 or 1. 
```{r check_cn}
table(mutations.bulk$totalCN)
```

Among 789x16 = 12624 entries in the mutation data matrix for the 789 loci and 16 single cells, 4562 ($\sim 36\%$) entries are missing due to low read-depth.  
```{r check_missing}
table(c(mutations.sc), useNA="ifany")
table(is.na(mutations.sc))/12624
```

Next we cluster the mutations based on the data from single cells. This involves imputing the missing values in single cell mutation calls. We evaluate the resutls with number of clusters being 2, 3, or 4. In the following resutls, ```scs$n2``` and ```scs$n3``` indicate results for 2 or 3 clusters, respectively.   
```{r scs}
scs = cluster.SCS(mutations.sc, asrec, trec, nclust=2:4)

names(scs)
scs$summary.cluster
scs$nclust.by.BIC

scs$n2$theta
scs$n3$theta
```

Next we perform joint anlaysis of bulk tumor data and single cell data. 

```{r basic}
b.cn    = mutations.bulk$totalCN
b.asrec = mutations.bulk$t_alt_count
b.trec  = mutations.bulk$t_alt_count + mutations.bulk$t_ref_count

cBSC = cluster.BSC(b.cn, b.asrec, b.trec, mutations.sc, asrec, trec, 
                   nclusters=2:4)

names(cBSC)
cBSC$summary.cluster
cBSC$nclust.by.BIC
```

Next we perform joint anlaysis of bulk tumor data and single cell data, without accouting for copy number information. To do this, we just set the copy number input to be 2 for all loci. 

```{r basic2}
b.cn  = rep(2, length(b.asrec))
cBSC2 = cluster.BSC(b.cn, b.asrec, b.trec, mutations.sc, asrec, trec, 
                    nclusters=2:4)

names(cBSC2)
cBSC2$summary.cluster
cBSC2$nclust.by.BIC
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r summary, fig.width = 8, fig.height = 2}

scs$n2$nu
cBSC$n2$nu
cBSC2$n3$nu

g0 = ggcorrplot(scs$n2$gamma, tl.cex=10, lab = FALSE, show.legend=F) + 
  ggtitle('SCS only') + theme(plot.title = element_text(size = 10))

g1 = ggcorrplot(cBSC$n2$gamma, tl.cex=10, lab = FALSE, show.legend=F) + 
  ggtitle("BaSiC w/ CNA") + theme(plot.title = element_text(size = 10))

g2 = ggcorrplot(cBSC2$n3$gamma, tl.cex=10, lab = FALSE, show.legend=F) + 
  ggtitle("BaSiC w/o CNA") + theme(plot.title = element_text(size = 10))

g0
g1
g2
```


Finally, we check mutation classification based on the three approaches. 

```{r summary2}
colSums(scs$n2$postPs)
colSums(cBSC$n2$postPs)
colSums(cBSC2$n3$postPs)

mut.scs = apply(scs$n2$postPs, 1, which.max)
mut.bsc = apply(cBSC$n2$postPs, 1, which.max)
mut.bsc.nocn = apply(cBSC2$n3$postPs, 1, which.max)

t2 = table(mut.bsc, mut.scs)
t3 = table(mut.bsc, mut.bsc.nocn)
t4 = table(mut.scs, mut.bsc.nocn)

ts = cbind(t2, t3)
ts
t4
```



```{r}
sessionInfo()
```
